### PR TITLE
Use snprintf in ToStringPrecisionLocale()

### PR DESCRIPTION
**Pull request by gpesquero**
Replacement of `std::ostringstream` with `std::snprintf` in function `ToStringPrecisionLocale()`, as proposed by @biodranik in [PR 8896](https://github.com/organicmaps/organicmaps/pull/8896), in order to obtain a better performance.

Columns:
* Old: `std::ostringstream`
* New: `std::snprintf`
* Ratio: New / Old * 100

Sample of values:
```
Old: 61269 ns, New:  8961 ns (Ratio 15 %)
Old: 65807 ns, New: 11539 ns (Ratio 18 %)
Old: 82231 ns, New: 10500 ns (Ratio 13 %)
Old: 37308 ns, New:  6847 ns (Ratio 18 %)
Old: 59884 ns, New:  9384 ns (Ratio 16 %)
Old: 66462 ns, New: 11885 ns (Ratio 18 %)
Old: 55923 ns, New:  8961 ns (Ratio 16 %)
Old: 60616 ns, New:  9500 ns (Ratio 16 %)
Old: 55731 ns, New:  7923 ns (Ratio 14 %)
Old: 25577 ns, New:  5039 ns (Ratio 20 %)
Old: 53923 ns, New:  8346 ns (Ratio 15 %)
Old: 37769 ns, New:  4384 ns (Ratio 12 %)
Old:  7962 ns, New:  2077 ns (Ratio 26 %)
Old: 36808 ns, New:  4039 ns (Ratio 11 %)
Old: 35538 ns, New:  4038 ns (Ratio 11 %)
Old:  9616 ns, New:  2231 ns (Ratio 23 %)
Old: 34269 ns, New:  3885 ns (Ratio 11 %)
Old: 36847 ns, New:  4384 ns (Ratio 12 %)
Old:  7153 ns, New:  1769 ns (Ratio 25 %)
Old: 36615 ns, New:  4269 ns (Ratio 12 %)
```
New implementation is faster, as it takes around 10%-25% of the time of the previous implementation.

### CODE REVIEW

File: platform/measurement_utils.cpp [hunk 1]
```diff
@@ -53,6 +74,43 @@ std::string ToStringPrecisionLocale(platform::Locale const & loc, double d, int
...
+{
+  #define BUF_SIZE 50
+
+  char formatBuf[BUF_SIZE];
+  switch(pr)
+  {
+    case 0: strcpy(formatBuf, "%0.0f"); break;
+    case 1: strcpy(formatBuf, "%0.1f"); break;
+    default: std::snprintf(formatBuf, BUF_SIZE, "%%0.%df", pr); break;
+  }
```
Comment (biodranik) [inline platform/measurement_utils.cpp:87]: If we're touching the performance... What about this?

```suggestion
  std::string formatBuf("%0.0f");
  if (pr > 0)
  {
    if (pr < 10)
      formatBuf[3] = '0' + pr;  // E.g. replace %0.0f with %0.1f
    else
      LOG(LERROR, ("Too big unsupported precision", pr));
  }
```

File: platform/measurement_utils.cpp [hunk 2]
```diff
@@ -53,6 +74,43 @@ std::string ToStringPrecisionLocale(platform::Locale const & loc, double d, int
...
+
+  char textBuf[BUF_SIZE];
+  std::snprintf(textBuf, BUF_SIZE, formatBuf, d);
```
Comment (biodranik) [inline platform/measurement_utils.cpp:90]:
```suggestion
  int const n = std::snprintf(textBuf, sizeof(textBuf), formatBuf.c_str(), d);
  if (n < 0 || n >= sizeof(textBuf))
  {
    LOG(LERROR, "snprintf", pr, d, "failed with", n);
    return std::to_string(static_cast<int64_t>(d));
  }
```

**Review by biodranik — COMMENTED** [review]
Thanks! Isn't it cool? )

**Review by biodranik — APPROVED** [review]
Thanks! LGTM, @vng PTAL

**Comment by gpesquero** [comment]
@rtsisyk: I've just figured out that with the previous implementation with `std::string`, we are able to manage UTF characters for decimal and grouping separators, such as NNBS (NarrowNonBreakingSpace = "\u202F").

But the plain C version that I just tested works Ok with regular chars, but not with UTF chars. We'd have to use `wchar_t`, convert it to `std::wstring` and then to `std::string` to make it work properly, which may complicate the code and we may not get any better performance.

So I'd leave the implementation as proposed, using `std::string` and no low-level C style functions.

**Comment by biodranik** [comment]
@gpesquero I'm curious, what was your C implementation?

**Comment by gpesquero** [comment]
> @gpesquero I'm curious, what was your C implementation?

Here's the piece of code: 
https://github.com/gpesquero/organicmaps/commit/e987628b32079840efdb44f456579d8133423a32#diff-0723736751723942b891b3be1579d306bfa25cfa0beaefeb809e77fb7b3e383bR133-R184

**Comment by biodranik** [comment]
Inserting a substring into a string is tricky in C, and obvious in C++.

**Review by biodranik — APPROVED** [review]

**Pull request merged**